lib/puppet/transaction.rb

```diff
@@ -148,13 +148,33 @@ def evaluate(&block)
       persistence.save if catalog.host_config?
     end
 
+    # Graph cycles are returned as an array of arrays
+    # - outer array is an array of cycles
+    # - each inner array is an array of resources involved in a cycle
+    # Short circuit resource evaluation if we detect cycle(s) in the graph. Mark
+    # each corresponding resource as failed in the report before we fail to
+    # ensure accurate reporting.
+    graph_cycle_handler = lambda do |cycles|
+      cycles.each do |cycle|
+        cycle.each do |resource|
+          # We add a failed resource event to the status to ensure accurate
+          # reporting through the event manager. Adding an event with status of
+          # 'failure' causes the status itself to be failed.
+          event = resource.event(:name => :resource_error, :status => "failure", :message => _('resource is part of a dependency cycle'))
+          resource_status(resource).add_event(event)
```

Comment:
Could we DRY this up so that we call a new method `Puppet::Resource::Status#failed("message")`, and then refactor `Status#failed_because` to call that method? As it is now, it looks like we're not calling `status.failed = true` above (on the status object as opposed to the newly created event for that status).